src/progress.rs

```diff
@@ -273,61 +461,168 @@ impl ProgressSet {
+                    return CandidacyStatus::Elected;
+                } else if next.has_quorum(&rejects) || self.configuration.has_quorum(&rejects) {
+                    return CandidacyStatus::Ineligible;
+                }
+            }
+            None => {
+                if self.configuration.has_quorum(&accepts) {
+                    return CandidacyStatus::Elected;
+                } else if self.configuration.has_quorum(&rejects) {
+                    return CandidacyStatus::Ineligible;
+                }
+            }
+        };
+
+        CandidacyStatus::Eligible
     }
 
     /// Determines if the current quorum is active according to the this raft node.
     /// Doing this will set the `recent_active` of each peer to false.
     ///
     /// This should only be called by the leader.
     pub fn quorum_recently_active(&mut self, perspective_of: u64) -> bool {
-        let mut active = 0;
+        let mut active = FxHashSet::default();
         for (&id, pr) in self.voters_mut() {
             if id == perspective_of {
-                active += 1;
+                active.insert(id);
                 continue;
             }
             if pr.recent_active {
-                active += 1;
+                active.insert(id);
             }
             pr.recent_active = false;
         }
         for (&_id, pr) in self.learners_mut() {
             pr.recent_active = false;
         }
-        active >= majority(self.voter_ids().len())
+        self.configuration.has_quorum(&active) &&
+            // If `next` is `None` we don't consider it, so just `true` it.
+            self.next_configuration.as_ref().map(|next| next.has_quorum(&active)).unwrap_or(true)
     }
 
     /// Determine if a quorum is formed from the given set of nodes.
+    ///
+    /// This is the only correct way to verify you have reached a quorum for the whole group.
+    #[inline]
     pub fn has_quorum(&self, potential_quorum: &FxHashSet<u64>) -> bool {
-        potential_quorum.len() >= majority(self.voter_ids().len())
+        self.configuration.has_quorum(potential_quorum)
+            && self
+                .next_configuration
+                .as_ref()
+                .map(|next| next.has_quorum(potential_quorum))
+                // If `next` is `None` we don't consider it, so just `true` it.
+                .unwrap_or(true)
+    }
+
+    /// Determine if the ProgressSet is represented by a transition state under Joint Consensus.
+    #[inline]
+    pub fn is_in_membership_change(&self) -> bool {
+        self.next_configuration.is_some()
+    }
+
+    /// Enter a joint consensus state to transition to the specified configuration.
+    ///
+    /// The `next` provided should be derived from the `ConfChange` message. `progress` is used as
+    /// a basis for created peer `Progress` values. You are only expected to set `ins` from the
+    /// `raft.max_inflights` value.
```

Comment:
It's not clear to me what `ins` represents here. It's an an argument to the function.